File: assets/codebase/rendering/RotateComponent.cs
namespace city.rendering {
    /// <summary>
    /// Applies a continuous authored local Y-axis rotation to the owning city showcase entity.
    /// </summary>
    public sealed class RotateComponent : UpdateComponent {
        /// <summary>
        /// Gets or sets the local radians advanced every frame update.
        /// </summary>
        public float RadiansPerFrame { get; set; } = 0.07f;

        /// <summary>
        /// Rotates the owning entity and normalizes the result to avoid drift.
        /// </summary>
        public override void Update() {
            base.Update();

            float4 deltaRotation;
            float3 axis = new float3(0f, 1f, 0f);
            float4.CreateFromAxisAngle(ref axis, RadiansPerFrame, out deltaRotation);

            float4 orientation = Parent.Orientation;
            orientation = deltaRotation * orientation;
            orientation.Normalize();

            Parent.Orientation = orientation;
        }
    }
}

File: windows-build-20260529-topple/cooked/code/gameplay/RotateComponent.cpp
#ifdef DrawText
#undef DrawText
#endif
#include "RotateComponent.hpp"
#include "UpdateComponent.hpp"
#include "float4.hpp"
#include "float3.hpp"
#include "Entity.hpp"

RotateComponent::RotateComponent() : RadiansPerFrame(0.07f)
{
}

float RotateComponent::get_RadiansPerFrame()
{
return this->RadiansPerFrame;
}

void RotateComponent::set_RadiansPerFrame(float value)
{
this->RadiansPerFrame = value;
}

void RotateComponent::Update()
{
UpdateComponent::Update();
float4 deltaRotation;
float3 axis = float3(0.0f, 1.0f, 0.0f);
float4::CreateFromAxisAngle(axis, this->RadiansPerFrame, deltaRotation);
float4 orientation = this->get_Parent()->get_Orientation();
orientation = deltaRotation * orientation;
orientation.Normalize();
Parent->set_Orientation(orientation);
}

